Additional changes to fully support Cancun

fselmo opened the pull request:
- Pull out the blob data when sending a transaction to the execution layer (py-evm). The execution layer does not deal with blob data, that's the whole point of blob transactions. We pull out the data and re-send with only the versioned hashes.

- Sum up the correct value for `blob_gas_used` during block proposal. This is validated in the execution layer (py-evm).

---

**Note: this will require: ethereum/py-evm#2165  and ethereum/py-evm#2166 to be merged and released. I've tested locally and these changes make it so we can send a blob transaction via *web3.py* -> `EthereumTesterProvider` -> *eth-tester*  -> *py-evm*.**


### How was it fixed?

- Remove the blob data and re-serialize blob transactions in a way that the EL expects, without the bloat of the blob data.
- Sum up `blob_gas_used` and send it to EL (py-evm) 
- Add validation and normalization for Cancun-related fields on block / header and transactions.


### Todo:

- [x] Clean up commit history
- [x] More testing + verify these are all of the remaining changes related to full Cancun support
- [x] Add entry to the [release notes](https://github.com/ethereum/eth-tester/blob/main/newsfragments/README.md)

#### Cute Animal Picture

<img width="722" alt="Screenshot 2024-04-03 at 17 54 20" src="https://github.com/ethereum/eth-tester/assets/3532824/f393b020-4b01-437a-9c9e-61d0e6a1e4d3">

fselmo committed:
Some final changes to support Cancun:

- Pull out the blob data when sending a transaction to the execution layer (py-evm).
  The execution layer does not deal with blob data, that's the whole point of blob
  transactions. We pull out the data and re-send with only the versioned hashes.

- Sum up the correct number of `blob_gas_used` since this is validated in the execution
  layer. This is basically simulating what the CL computes and sends to EL.

 - Add validation and normalization for Cancun-related fields for blocks
  / block headers and transactions

- Add `y_parity` to non-legacy receipts

- newsfragment for PR #285

fselmo committed:
Add some more basic tests around cancun field validation

fselmo committed:
Keep ``py-evm`` within the Cancun minor beta version.

fselmo committed:
Add some basic tests for sending raw blob transactions

kclowes commented on eth_tester/normalization/common.py:
```diff
@@ -2,7 +2,7 @@
     is_hex,
     to_dict,
-    to_tuple,
+    to_list,
```
curious about this change?

fselmo replied:
Consistent with output from clients. They are usually lists, not tuples in the JSON.

fselmo commented on eth_tester/backends/mock/serializers.py:
```diff
@@ -52,6 +52,13 @@ def serialize_full_transaction(transaction, block, transaction_index, is_pending
 
+    if int(serialized_transaction["type"], 16) > 0:
+        serialized_transaction = assoc(
+            serialized_transaction,
```
This was a bonus that I noticed clients return. There is still a `v` field for non-legacy transactions but there's also the correct `y_parity` field of the signature.

reedsa commented on eth_tester/backends/pyevm/main.py:
```diff
@@ -734,7 +750,24 @@ def _create_type_aware_unsigned_transaction(self, normalized_txn):
         vm = _get_vm_for_block_number(self.chain, "latest")
-        evm_transaction = vm.get_transaction_builder().decode(raw_transaction)
+
+        if raw_transaction[0] == 3:
```
Nit: use a constant to make it easy to tell what the number 3 means.

reedsa commented on eth_tester/backends/pyevm/serializers.py:
```diff
@@ -74,13 +77,30 @@ def serialize_block(block, full_transaction, is_pending):
         block_info.update({"withdrawals": serialize_block_withdrawals(block)})
         block_info.update({"withdrawals_root": block.header.withdrawals_root})
 
+    if all(
```
Similar to other comment, I wonder if this check is useful elsewhere. If so it might be worth calling a boolean function with a descriptive name.

Otherwise I'm good with the comment you used in this block.

reedsa commented on tests/backends/test_pyevm.py:
```diff
@@ -432,3 +443,54 @@ def test_eth_get_storage_at(self):
+
+    # --- cancun network upgrade --- #
+
+    BLOB_TEXT = "We are the music makers, And we are the dreamers of dreams."
```
:smile:

reedsa commented on eth_tester/backends/pyevm/serializers.py:
```diff
@@ -103,13 +123,8 @@ def serialize_transaction(block, transaction, transaction_index, is_pending):
-        else transaction.y_parity,
     }
-    if _field_in_transaction(transaction, "gas_price"):
+    if int(txn_type, 16) in (0, 1):
```
Moar nit: Worth a constant here as well? As a naive person reading I don't know offhand what `txn_type` being 0 or 1 means.

Comment or descriptive name for this could help.

reedsa reviewed (approved):
Nice work, I left a couple comments you can take or leave. Not a blocker in any sense.

kclowes commented on eth_tester/validation/outbound.py:
```diff
@@ -277,6 +312,19 @@ def _validate_fork_specific_fields(block):
 
+    # Cancun fork
+    if all(
+        _ not in block
```
I like @reedsa's suggestion to make these into methods. I think it would be easier to reason about.

kclowes reviewed (approved):
Looks good to me! I put this in the comments too, but I like @reedsa's idea to have descriptive method names instead of `if any(...)` and `if all(...)`

pacrob reviewed (approved):
Looks good! I like the Stu idea of using constants for transaction types for readability. Whaddya think?

fselmo committed:
Changes from comments on PR #285:

- Refactor block evm guessing logic into utility methods
- Use constants for transaction types for clarity
- Keep integer for transaction type internally, normalize to hex
  (as clients do) on the way out.

reedsa reviewed (approved):
Looks awesome, great refactor!
